File: z_lc/5-longest-palindromic-substring/solution-1-dp/longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        
        bool memo[n][n];
        
        int start = 0, len = 0;
        for(int i=n-1; i>=0; --i) {
            for(int j=i; j<n; ++j) {
                if(i == j) {
                    memo[i][j] = true;
                } else if(i+1 == j) {
                    memo[i][j] = s[i] == s[j];
                } else {
                    memo[i][j] = (s[i] == s[j]) && memo[i+1][j-1];
                }
                if(memo[i][j] && len < j - i + 1) {
                    start = i;
                    len = j - i + 1;
                }
            }
        }
        return s.substr(start, len);
    }
};

File: z_lc/5-longest-palindromic-substring/solution-1-dp/longest-palindromic-substring.rs
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let S = s.as_bytes();
        let n = S.len();
        
        let mut memo = vec![vec![false; n]; n];
        
        let (mut start, mut len) = (0, 0);
        for i in (0..n).rev() {
            for j in i..n {
                memo[i][j] = match (i,j) {
                    (a, b) if a == b => true,
                    (a, b) if a+1 == b => S[i] == S[j],
                    _ => (S[i] == S[j]) && memo[i+1][j-1],
                };
                if memo[i][j] && len < j - i + 1 {
                    start = i;
                    len = j - i + 1;
                };
            }
        }
        s[start..start+len].to_string()
    }
}